Analysis/bin/CreateSpectralHists.cxx

```diff
@@ -0,0 +1,145 @@
+*/
+#include "TauMLTools/Core/interface/program_main.h"
+#include "TauMLTools/Analysis/interface/TauTuple.h"
+#include "TauMLTools/Analysis/interface/SummaryTuple.h"
+#include "TauMLTools/Core/interface/RootFilesMerger.h"
+#include "TauMLTools/Core/interface/NumericPrimitives.h"
+
+#include "TauMLTools/Core/interface/AnalyzerData.h"
+#include "TauMLTools/Core/interface/RootExt.h"
+
+struct Arguments {
+    run::Argument<std::string> outputfile{"outputfile", "output file name"};
+    run::Argument<std::vector<std::string>> input_dirs{"input-dir", "input directory"};
+    run::Argument<std::string> pt_hist{"pt-hist", "pt hist setup: (number of bins, pt_min, pt_max)", "200, 0.0, 1000"};
+    run::Argument<std::string> eta_hist{"eta-hist", "eta hist setup: (number of bins, abs(eta)_min, abs(eta)_max)","4, 0.0, 2.3"};
+    run::Argument<std::string> file_name_pattern{"file-name-pattern", "regex expression to match file names",
+                                                 "^.*\\.root$"};
+    run::Argument<std::string> exclude_list{"exclude-list", "comma separated list of files to exclude", ""};
+    run::Argument<std::string> exclude_dir_list{"exclude-dir-list",
+                                               "comma separated list of directories to exclude", ""};
+};
+
+struct HistArgs {
+  int eta_bins, pt_bins;
+  double eta_min, eta_max, pt_min, pt_max;
+  HistArgs(std::vector<std::string> args_pt, std::vector<std::string> args_eta)
+  {
+    pt_bins = analysis::Parse<int>(args_pt[0]);
+    pt_min = analysis::Parse<double>(args_pt[1]);
+    pt_max = analysis::Parse<double>(args_pt[2]);
+    eta_bins = analysis::Parse<int>(args_eta[0]);
+    eta_min = analysis::Parse<double>(args_eta[1]);
+    eta_max = analysis::Parse<double>(args_eta[2]);
+  }
+};
+
+class HistSpectrum : public root_ext::AnalyzerData {
+
+public:
+  HistSpectrum(std::shared_ptr<TFile> outputFile, HistArgs hargs) : root_ext::AnalyzerData(outputFile)
+  {
+    eta_pt_hist.SetMasterHist(hargs.eta_bins, hargs.eta_min, hargs.eta_max,
+                              hargs.pt_bins, hargs.pt_min, hargs.pt_max);
+  }
+
+  TH2D_ENTRY(eta_pt_hist, 4, 0, 2.3, 200, 0, 1000)
+
+};
+
+namespace analysis {
+class CreateSpectralHists {
+
+private:
+    std::vector<std::string> input_files;
+    std::shared_ptr<TFile> outputfile;
+    std::shared_ptr<HistSpectrum> hists;
+    Int_t total_size=0;
+
+public:
+    using Tau = tau_tuple::Tau;
+    using TauTuple = tau_tuple::TauTuple;
+
+    CreateSpectralHists(const Arguments& args) :
+      input_files(RootFilesMerger::FindInputFiles(args.input_dirs(),
+                                                  args.file_name_pattern(),
+                                                  args.exclude_list(),
+                                                  args.exclude_dir_list())),
+       outputfile(root_ext::CreateRootFile(args.outputfile()))
+    {
+      auto par_path = GetPathWithoutFileName(args.outputfile());
```

Comment:
To speed-up the decompression during reading (which takes most of the time), you can enable an implicit multithreading in ROOT:
```c++
ROOT::EnableThreadSafety();
if(args.n_threads() > 1)
    ROOT::EnableImplicitMT(args.n_threads());
```

Reply:
@kandrosov enabling only branches that I need reduced time from 40mins to ~5-6mins, but for some reasons ROOT::EnableImplicitMT with more than one thread slows down everything significantly,

Reply:
Interesting. Have you tested it several times (or on a machine where you fully control the conditions)? For lxplus, for example, from one run to another it can happen that someone else started a heavy job and everything goes slower...

However, it can be the case that the amount of enabled branches is so small, so there is no gain from parallelizing the decompression.

Please, choose the fastest option. Below 10 minutes per a large dataset (as W2JetsToLNu) is acceptable - we need to create spectrum histograms only once per production.

Reply:
Yes, I've tested it many times, the time increases very significantly, I am testing it on desy naf-cms.desy.de the machine is not loaded at all, if I run let's say with --n-threads 6 it loads full 6 cores

Reply:
Ok, thanks for the details! I propose then to keep running the code with a single thread and move forward.
Creating binned spectrum is not a recurrent task: it enough to create histograms with fine binning (e.g. 1 GeV step for pt and 0.01 for eta) for each dataset once and then rebin them according to the needs when running ShuffleMerge. Therefore, as long as the total time to produce a spectrum for the all datasets is O(1 day) - it is acceptable.

Maybe you can also add histograms with the pt-eta spectrum of seeding jets - it is not needed for the current training, but can be used in the future (for example, for an ML-based tau reco training).